### PR TITLE
Implement ``raster_reduce``

### LINKED ISSUE #285: Proposal: `raster_reduce` for raster operations that don't produce a raster
It could be useful to have an equivalent of `raster_calculator` for operations that don't produce a raster output. If `raster_calculator` is like `map` over the blocks of a raster, then this would be like `functools.reduce` over the blocks of a raster.

Aggregations are probably the most common use case. For the Morgan Stanley project I needed to produce histogram data for a raster. `raster_calculator` seems like the easiest approach, but it requires writing to a target raster which was totally irrelevant for my use case. I ended up with:
```python
def raster_histogram(raster_path, bins=10):
    """Return histogram data for the given raster divided into bins."""

    histogram = numpy.zeros(bins)

    def add_to_histogram(array):
        nonlocal histogram
        histogram += numpy.histogram(array, bins=bins)[0]
        return array

    pygeoprocessing.raster_calculator(
        [(raster_path, 1)],
        add_to_histogram,
        'foo.tif',
        gdal.GDT_Int32,
        TARGET_NODATA
    )
    return histogram
```

Given a proposed new function `raster_reduce(raster_path, function, initializer)` that behaves like [`functools.reduce`](https://docs.python.org/3/library/functools.html#functools.reduce), that could be rewritten as

```python
def raster_histogram(raster_path, bins=10):
    """Return histogram data for the given raster divided into bins."""

    def add_to_histogram(array, histogram):
        return histogram + numpy.histogram(array, bins=bins)[0]

    return pygeoprocessing.raster_reduce(
        [(raster_path, 1)],
        add_to_histogram,
        numpy.zeros(bins))
```

I'm not sure if there are any use cases for this function in invest, but I imagine it would be useful for data analysis.

### PR DESCRIPTION
**Pull request by emlys**
Fixes #285 
I didn't include any error handling like `raster_calculator` has, so the implementation is very simple. Let me know if you think it's missing anything!

### CODE REVIEW

File: src/pygeoprocessing/geoprocessing.py
Comment (phargogh) [inline src/pygeoprocessing/geoprocessing.py]: I can see the parallels to python's reduce, which is neat!  A couple thoughts:

1. What would you think about using the `(path, band)` notation, for consistency with other raster operations in this file?
2. Would it be worth having the path/band tuple as the first parameter?  I see that most functions in `pygeoprocessing` that involve rasters do this, but also the function as it stands is more like the stdlib reduce.  What do you think?
3. One thing that's interesting and convenient about the `functools` reduce is that the `initializer` is optional.  Would that be worth including here as well?
4. Since this is a wrapper around `iterblocks`, I wonder if it would be helpful to also expose the `largest_block` parameter, in case the user knows something about their system and wants to fiddle with the block size.

File: src/pygeoprocessing/geoprocessing.py
Comment (phargogh) [inline src/pygeoprocessing/geoprocessing.py]: Thinking about `raster_calculator` and how long it can take to iterate through potentially very large rasters, maybe some timed progress logging would be helpful here?

File: tests/test_geoprocessing.py
Comment (phargogh) [inline tests/test_geoprocessing.py]: Since this is derived from `array`, maybe we could just use `numpy.sum(array)` to make it clear that we're just summing pixel values?

File: tests/test_geoprocessing.py
Comment (phargogh) [inline tests/test_geoprocessing.py:4873]: I think I get what this is doing, but maybe a comment would help make the intent clear?

File: tests/test_geoprocessing.py
Comment (phargogh) [inline tests/test_geoprocessing.py]: Bonus lines?

**Review by phargogh — COMMENTED** [review]
I mostly just have a bunch of questions/comments about the API and a couple other things, let me know what you think!

File: tests/test_geoprocessing.py
Comment (emlys) [reply tests/test_geoprocessing.py]: good idea!

File: src/pygeoprocessing/geoprocessing.py
Comment (emlys) [reply src/pygeoprocessing/geoprocessing.py]: 1: Sure!
2: I did put the arguments in this order to resemble `functools.reduce`, but consistency with pygeoprocessing is an equally valid reson. I don't really have a preference either way. 
4: Good idea!

Comment (emlys) [reply src/pygeoprocessing/geoprocessing.py]: 3: I think the `initializer` is necessary with the current design, but a maybe not with a different design.

With `functools.reduce(function, iterable, initializer)`, the initializer is optional because you can infer it. If `initializer` isn't provided, it's set to the return value of the first `function` call. 

But `raster_reduce` has two parts: it aggregates each block to a single value and then reduces _those_ values to a single value. In the current implementation, the `function` does both of those operations. For instance with the histogram example, the correct value of `initializer` to infer would be the result of `numpy.histogram` called on the first block. But there's no way to know that because `numpy.histogram` is wrapped up in the `function`.

But it would be possible if we split out the block aggregation and reducing operations into separate functions, something like:
```
def raster_reduce(raster_path, block_aggregate_fn, raster_reduce_fn, initializer=None):
    # reduce each block to a single value
    block_values = map(block_aggregate_fn, iterblocks(raster_path))
    if initializer is not None:
        block_values = [initializer] + block_values
    # reduce that series of block values to a single value
    return functools.reduce(raster_reduce_fn , block_values)
```
Then the histogram example could be written as
```
def block_histogram(block): return numpy.histogram(block, bins=10)[0]
def sum(x, y): return x + y
raster_reduce(raster_path, block_histogram, sum)
```
This is kind of hard to wrap my head around, does that make any sense? And would that alternative implementation be easier to use?

Comment (phargogh) [reply src/pygeoprocessing/geoprocessing.py]: Aha!  Yes, you're absolutely right.  Thanks for walking me through this.  The return type of `function` might not be an array the size and shape (or even type) of `block`, so in general we can't just take the first block and use that as the initializer.

Ah, higher-order functions are so interesting!  I think we should stick with your original formulation with the initializer so that we minimize the number of functions one has to understand in order to get this working in code.  Conceptually the combination of map and reduce in the alternate implementation makes an interesting distinction, but I think the original implementation might be more straightforward.

Comment (phargogh) [reply src/pygeoprocessing/geoprocessing.py]: I had one other thought that I wonder what you think about:  would it make sense or be useful to optionally filter out nodata values?  We could always write our reduce function to handle nodata, but that requires knowing the nodata ahead of time when it could easily be read and the block indexed within `raster_reduce`.  If you agree, would it also make sense to flatten out the block for cases where we are not doing nodata filtering?  Just a thought!

File: src/pygeoprocessing/geoprocessing.py
Comment (phargogh) [inline src/pygeoprocessing/geoprocessing.py]: Such a minor thing, but slightly cheaper :)

```suggestion
        pixels_processed += block.size
```

**Review by phargogh — CHANGES_REQUESTED** [review]
This is looking really cool @emlys !  And thanks for walking me through the need for the initializer, that totally makes sense that we need it.  I just had one trivial tweak to suggest and a question about whether it'd make sense to handle nodata ... let me know what you think!

File: src/pygeoprocessing/geoprocessing.py
Comment (emlys) [reply src/pygeoprocessing/geoprocessing.py]: Good idea!

**Comment by emlys** [comment]
@phargogh I added a nodata masking option like you suggested, and made that the default. For consistency I also `flatten`ed the block if `mask_nodata` is False. Let me know what you think!

**Review by phargogh — APPROVED** [review]
Looks great!  I agree, flattening for consistency sounds right to me.

Just waiting until tests pass to merge!

**Comment by emlys** [comment]
Merging now that tests passed

**Comment by emlys** [comment]
Github is glitching here... the merge failed even though there were no conflicts, so I merged locally, then it marked the PR as closed.

**Pull request closed without merging**
